### PR TITLE
Make AFQDataset work with different input formats and make indexable

### LINKED ISSUE #98: Add capabilities to split and impute AFQDataset
Dataset splitting and imputation is common and has the potential to introduce train/test leakage if done incorrectly. It's also easier to do when all of your dataset is in memory (i.e. before being batched into tensorflow or pytorch datasets.

## Proposed solution

- [ ] Make `AFQDataset` initialization more general so that user can provide `X`, `y`, `groups`, `feature_names`, `group_names`, `subjects`, `sessions`, `classes` explicitly on init.
- [ ] create a new static method `from_files()` that takes all of the filenames that are in the current `AFQDataset` init method and returns an `AFQDataset` object with all of the data read in.
- [ ] create a new `split()` method that will apply an [sklearn splitter class](https://scikit-learn.org/stable/modules/classes.html#splitter-classes) and wrap its `split` method (which usually returns tuples of indices) by returning a tuple of `AFQDataset` objects representing the train and test splits. The user is free to recursively split the test set if they want both test and validate splits.
- [ ] create a `train_test_split()` that wraps `next(split(ShuffleSplit()))` in much the same way that `train_test_split` works in sklearn.

## Topics for discussion

We often need to impute missing bundles and have to take care to fit imputation on only the training set to avoid data leakage. It would be nice to provide the user with an option to do this during splitting. One option is to add `imputer` and `imputer_kwargs` parameters to the aforesaid `split()` method and impute within the wrapper for the splitters split method. I like this option but am open to other implementations.

Tagging @arokem to continue a discussion that we had about this in person. Note my change of heart here in making `AFQDataset` more general and reading from input files using a static method.

### PR DESCRIPTION
**Pull request by richford**
Resolves #98
This PR
- [x] makes AFQDataset initialization more general so that user can provide X, y, groups, feature_names, group_names, subjects, sessions, classes explicitly on init.
- [x] create a new static method from_files() that takes all of the filenames that are in the current AFQDataset init method and returns an AFQDataset object with all of the data read in.
- [x] Make AFQDataset more array-like by adding `__getitem__` and `__len__` methods and a shape parameter. This means users can do things like `dataset[10:20]` and also that `AFQDataset`s can be used as input to sklearn functions like `train_test_split`.
- [x] Adds a lot of detail and doctests to `AFQDataset`'s docstring

Note that I chose a different solution to dataset splitting than the one I proposed in #98. I think making the datasets indexable and therefore interoperable with scikit-learn's already existing, performant, and robust model selection routines is a much better solution than rolling our own `split` method.

This doesn't resolve the need for imputation methods, which we hint at in #98 but I now think that if we still want those, we should open up another issue and PR for that.

### CODE REVIEW

File: afqinsight/datasets.py [hunk 1]
```diff
@@ -477,18 +571,68 @@ def __init__(
             enforce_sub_prefix=enforce_sub_prefix,
         )
 
-        self.X = afq_data.X
-        self.y = afq_data.y
-        if self.y is not None:
-            self.y = self.y.astype(float)
+        y = afq_data.y
+        if afq_data.y is not None:
+            y = afq_data.y.astype(float)
```
Comment (arokem) [inline afqinsight/datasets.py:576]: Why is this necessary?

File: afqinsight/tests/test_datasets.py [hunk 2]
```diff
@@ -133,10 +133,42 @@ def test_afqdataset_sub_prefix():
...
+    assert len(dataset[:2]) == 2
+    assert isinstance(dataset[:2], AFQDataset)
+    assert repr(dataset) == "AFQDataset(n_samples=10, n_features=4, targets=None)"
+
+    dataset = AFQDataset(X=np.random.rand(10, 4))
+    assert len(dataset) == 10
+    assert dataset.shape == (10, 4)
+    assert len(dataset[:2]) == 2
+    assert isinstance(dataset[:2], AFQDataset)
+    assert repr(dataset) == "AFQDataset(n_samples=10, n_features=4, targets=None)"
```
Comment (arokem) [inline afqinsight/tests/test_datasets.py:149]: Should we make a different repr for the unsupervised case? For example, drop the `targets=None` from the repr in that case?

**Review by arokem — COMMENTED** [review]
Overall, looks great! I think that a sphinx gallery example of using `train_test_split` with some real data should be helpful. But we can take that on a separate PR.

File: afqinsight/datasets.py (hunk 1 above)
Comment (richford) [reply afqinsight/datasets.py]: It's not! It's necessary when transforming to a tf or pytorch dataset but I moved that logic into those methods. Thanks for catching this!

File: afqinsight/tests/test_datasets.py (hunk 2 above)
Comment (richford) [reply afqinsight/tests/test_datasets.py]: Good suggestion. Implemented in the latest commit.

File: afqinsight/datasets.py (hunk 1 above)
Comment (richford) [reply afqinsight/datasets.py]: Hmm. Or maybe not given the recent CI failure. I'll investigate and update.

Comment (richford) [reply afqinsight/datasets.py]: Okay, now I remember why that was there. I wanted to cast to float so that it could accommodate NaNs in the unit tests, which now strikes me as backwards. It should be up to the user to manipulate their data so that it can accommodate NaNs if needed. Fixed in the latest commit.
